File: imperium-common/src/main/java/com/xpdustry/imperium/common/database/SQLDatabase.java
package com.xpdustry.imperium.common.database;

import com.xpdustry.imperium.common.functional.ThrowingConsumer;
import com.xpdustry.imperium.common.functional.ThrowingFunction;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.Instant;
import java.util.stream.Stream;

public interface SQLDatabase {

    <R> R withFunctionHandle(final ThrowingFunction<Handle, R, SQLException> function);

    default void withConsumerHandle(final ThrowingConsumer<Handle, SQLException> consumer) {
        this.<Void>withFunctionHandle(handle -> {
            consumer.accept(handle);
            return null;
        });
    }

    void executeScript(final String script);

    interface Handle {

        StatementBuilder prepareStatement(final String statement);
    }

    interface StatementBuilder {

        StatementBuilder push(final String value);

        StatementBuilder push(final int value);

        StatementBuilder push(final long value);

        StatementBuilder push(final boolean value);

        StatementBuilder push(final float value);

        StatementBuilder push(final byte[] value);

        StatementBuilder push(final Instant value);

        <T> Stream<T> executeSelect(final ThrowingFunction<ResultSet, T, SQLException> mapper);

        int executeUpdate();

        boolean executeSingleUpdate();
    }
}

File: imperium-common/src/main/kotlin/com/xpdustry/imperium/common/database/SQLDatabase.kt
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.imperium.common.database

import java.time.Instant

interface SQLDatabase {
    suspend fun <R> transaction(block: suspend Handle.() -> R): R

    interface Handle {
        fun String.asPreparedStatement(): PreparedStatementBuilder
    }

    interface PreparedStatementBuilder {
        fun push(value: String): PreparedStatementBuilder

        fun push(value: Int): PreparedStatementBuilder

        fun push(value: Long): PreparedStatementBuilder

        fun push(value: Boolean): PreparedStatementBuilder

        fun push(value: Double): PreparedStatementBuilder

        fun push(value: Float): PreparedStatementBuilder

        fun push(value: ByteArray): PreparedStatementBuilder

        fun push(value: Instant): PreparedStatementBuilder

        // TODO pushRow or rename that whole shi...

        suspend fun <T> executeSelect(mapper: Row.() -> T): MutableList<T>

        suspend fun executeUpdate(): Int

        suspend fun executeSingleUpdate(): Boolean
    }

    interface Row {
        fun getString(name: String): String?

        fun getInt(name: String): Int?

        fun getLong(name: String): Long?

        fun getBoolean(name: String): Boolean?

        fun getDouble(name: String): Double?

        fun getFloat(name: String): Float?

        fun getBytes(name: String): ByteArray?

        fun getInstant(name: String): Instant?
    }
}
